### ISSUE TITLE
Some questions when reproducing the results in the DSNAS paper

### ISSUE THREAD

**Issue by zhangzjn** [SNAS-Series/SNAS-Series#4]
Dear Shoukang,
thanks for your great job and I can reproduce the results in the DSNAS paper following the provided steps. But I have some questions about the codes, as listed below:
1. When directly training the DSNAS for 240 epochs, I can only get `R1@71.584` that has a gap with the final result (R1@74.4).
2. Why you suggest early-stop epocs equaling 80? That  does not agree with the formula-(8) in the paper.
3. How do you set the threshold h (in formula-(8)) in the paper? And I can't obtain the `log_alpha` closed to one-hot vectors, e.g. [0.5588, 0.2505, 0.1286, 0.0622], [0.3196, 0.2729, 0.3506, 0.0570].
4. The second step that continuing the searching stage from epoch 80 seems to only finetune the fixed architecture obtained from `weights = torch.zeros_like(log_alpha).scatter_(1, torch.argmax(log_alpha, dim = -1).view(-1,1), 1)` in `train_imagenet_child.py` and the `log_alpha ` will not change, which violates the concept of end-to-end.

Looking forward to your replies. 
Thanks a lot!

**Comment by skhu101** [comment]
Thanks for your interest in our paper. 
1. I think there are two main reasons behind the R1@71.584. 1) R1@71.584 is the result without doing BN recalibration since the statistics in BN are not accurate. 2) When we train the supernet, we find that choice blocks in a few layers do not satisfy the formula-(8) and it keeps sampling the chocie blocks during the training. So these two factors will deterioterate the performance a bit.
2. Acorrding to the alpha curve, we find most layers satisfy the formula-(8) in epoch 80, so we set the early-stop epochs to be 80. Anyway, it is an empirical number and you can tune this. 
3. We set the h to be 0.3. If the difference between the highest probability and second largest probability in one layer is larger than 0.3, we apply the early-stopping stategy.
4. In the first step, we search the superent with the early-stopping strategy. Then in the second step we fix all the layers to be the choice block with the highest probability and continue tuning the derived models. End-to-end in our paper is defined to be co-optimizing parameter and architecture in one optimization process, in contrast to parameter retraining as done in other NAS pipelines. Actually, even we continue sampling the choice blocks in the second step of our experiments, the relative order of alpha values do not change so much, so we directly fix the alpha values.

**Comment by zhangzjn** [comment]
@skhu101 
Thanks for your helpful reply!
